pkg/metadata/manager.go

```diff
@@ -0,0 +1,374 @@
+}
+
+type containerMetadata struct {
+	cgroupId    uint64
+	cgroupPath  string
+	parentPodId string
+}
+
+type podEvent struct {
+	pod            *v1.Pod
+	oldPod         *v1.Pod
+	cgroupId       uint64
+	eventType      int
+	podMetadataOut chan *PodMetadata
+}
+
+type fs interface {
+	Stat(name string) (os.FileInfo, error)
+}
+
+type cgroup interface {
+	GetCgroupID(cgroupPath string) (uint64, error)
+}
+
+type fsImpl struct{}
+
+func (f fsImpl) Stat(name string) (os.FileInfo, error) {
+	return os.Stat(name)
+}
+
+type cgroupImpl struct{}
+
+func (c cgroupImpl) GetCgroupID(cgroupPath string) (uint64, error) {
+	return cgroups.GetCgroupID(cgroupPath)
+}
+
+func (m *CgroupManager) enable() {
+	go m.processPodEvents()
+	if !option.Config.EnableSocketLBTracing {
+		m.enabled = false
+		return
+	}
+	m.enabled = true
+	m.checkCgroupPath.Do(func() {
+		for _, path := range cgroupBasePaths {
+			if _, err := m.fschecker.Stat(path); err != nil {
+				continue
+			}
+			m.templateCgroupBasePath = path
+			break
+		}
+		if m.templateCgroupBasePath == "" {
+			log.Warnf("No valid cgroup base path found: socket " +
+				"load-balancing tracing feature will not work")
+			m.enabled = false
+		}
+	})
+
+	if m.enabled {
+		log.Info("Metadata manager is enabled")
+	}
+}
+
+func (m *CgroupManager) processPodEvents() {
+	for {
+		select {
+		case ev := <-m.podEvents:
+			if !m.enabled {
+				continue
+			}
+			switch ev.eventType {
+			case PodAddEvent, PodUpdateEvent:
+				m.updatePodMetadata(ev.pod, ev.oldPod)
+			case PodDeleteEvent:
+				m.deletePodMetadata(ev.pod)
+			case PodMetadataEvent:
+				m.getParentPodMetadata(ev.cgroupId, ev.podMetadataOut)
+			}
+		case <-m.shutdown:
+			return
+		}
+	}
+}
+
+func (m *CgroupManager) updatePodMetadata(pod, oldPod *v1.Pod) {
+	id := string(pod.ObjectMeta.UID)
+	pm, ok := m.podMetadataById[id]
+	if !ok {
+		// Fill in pod static metadata.
+		pm = &podMetadata{
+			// TODO: remove id?
+			id:        id,
+			name:      pod.Name,
+			namespace: pod.Namespace,
+		}
+		m.podMetadataById[id] = pm
+	}
+	if oldPod != nil && oldPod.Status.DeepEqual(&pod.Status) {
+		return
+	}
+	// Only update the metadata that can change. This excludes pod's name,
+	// namespace, id, and qos class.
+	podIPs := pod.Status.PodIPs
+	pm.ips = make([]string, len(podIPs))
+	for i := range podIPs {
+		pm.ips[i] = podIPs[i].IP
+	}
+	// Get metadata for pod's containers that are in the running state. Containers
+	// can get re-created, and their ids can change. Update the new containers.
+	// Pod's metadata including its containers map will be deleted when the pod
+	// is deleted.
+	numContainers := len(pod.Status.ContainerStatuses)
+	if pm.containers == nil && numContainers > 0 {
+		pm.containers = make(map[string]struct{})
+	}
+	currContainers := make(map[string]struct{}, numContainers)
+	for _, c := range pod.Status.ContainerStatuses {
+		var cId string
+		if cId = c.ContainerID; cId == "" || c.State.Running == nil {
```

Comment:
If one container in a pod crashes and is not restarted, should we delete it from the store?

Reply:
Nevermind :-)